proto/Peer.proto

```diff
@@ -2,15 +2,106 @@ syntax = "proto3";
 
 package peer;
 
-message HandshakeMessage {
-  bytes target_pub_key = 1;
-  bytes requesting_pub_key = 2;
-  bytes message = 3;
-  bytes response_peer_info = 4;
+//////////////////
+// Peer Service //
+//////////////////
+service Peer {
+  rpc MessagePeer (PeerMessage) returns (PeerMessage) {};
+}
+message PeerMessage {
+  string public_key = 1;
+  SubServiceType type = 2;
+  string sub_message = 3;
+}
+enum SubServiceType {
+  PING_PEER = 0;
+  GIT = 1;
+  NAT_TRAVERSAL = 2;
 }
 
+///////////////
+// Ping Peer //
+///////////////
+message PingPeerMessage {
+  string public_key = 1;
+  string challenge = 2;
+  PeerInfoMessage peerInfo = 3;
+}
 message PeerInfoMessage {
-  string pub_key = 1;
-  repeated string addresses = 2;
-  string connected_addr = 3;
+  string public_key = 1;
+  string connected_address = 2;
+  string relay_public_key = 3;
+}
+message PeerInfoListMessage {
+  repeated PeerInfoMessage peer_info_list = 1;
+}
+
+///////////
+// Error //
+///////////
+message ErrorMessage {
+  string error = 1;
+}
+
+///////////////////
+// NAT Traversal //
+///////////////////
+enum NatMessageType {
+  ERROR = 0;
+  RELAY_CONNECTION = 1;
+  PEER_CONNECTION = 2;
+  UDP_ADDRESS = 3;
+  PEER_UDP_ADDRESS = 4;
+}
+message NatMessage {
+  NatMessageType type = 1;
+  bool is_response = 2;
+  bytes sub_message = 3;
+}
+
+//////////////////////
+// Relay Connection //
+//////////////////////
+message RelayConnectionRequest {
+  string public_key = 1;
+}
+message RelayConnectionResponse {
+  string server_address = 1;
+}
+
+/////////////////////
+// Peer Connection //
+/////////////////////
+message PeerConnectionRequest {
+  string public_key = 1;
+}
+message PeerConnectionResponse {
+  string connected_address = 1;
+}
+
+/////////////////
+// UDP Address //
+/////////////////
+message UDPAddressResponse {
+  string address = 1;
+}
+
+/////////////////////////
+// Hole Punch Register //
+/////////////////////////
+message HolePunchRegisterRequest {
+  string public_key = 1;
+}
+message HolePunchRegisterResponse {
+  string connected_address = 1;
+}
+
+//////////////////////
+// Peer UDP Address //
+//////////////////////
+message PeerUdpAddressRequest {
```

Comment:
Do you have documentation on all your GRPC messages? In the future, any PRs to the protocbuf/grpc stuff should have on the MR description links to the API schema/API docs for the API.

Like for HTTP, that's swagger openapi.yaml file.

Is there an equivalent for gRPC?

The docs should indicate an almost like flow like diagram.

RPC calls are like functions to a library, should be documented similarly. A bit more effort because public API should be documented more strongly than private API.

Reply:
yeah it looks like there is [protoc-gen-doc](https://github.com/sourcegraph/prototools/blob/master/README.doc.md) but this uses Go HTML templates so we would have to have go installed in our shell. Then there is [protoc-gen-doc-node](https://www.npmjs.com/package/protoc-gen-doc-node) but I wouldn't trust it as it hasn't been updated in 2 years and just looks like it installs the go tools anyway